src/uu/du/src/du.rs

```diff
@@ -255,15 +296,310 @@ fn read_block_size(s: Option<&str>) -> UResult<u64> {
+                    // If we can't get std metadata, create a minimal fake one
+                    // This should rarely happen but provides a fallback
+                    fs::symlink_metadata("/").expect("root should be accessible")
+                });
+
+                Stat {
+                    path: path.to_path_buf(),
+                    size: if safe_metadata.is_dir() {
+                        0
+                    } else {
+                        safe_metadata.len()
+                    },
+                    blocks,
+                    inodes: 1,
+                    inode: file_info_option,
+                    metadata: std_metadata,
+                }
+            }
+            Err(e) => {
+                let error = e.map_err_context(
+                    || translate!("du-error-cannot-access", "path" => path.quote()),
+                );
+                if let Err(send_error) = print_tx.send(Err(error)) {
+                    return Err(Box::new(send_error));
+                }
+                return Err(Box::new(mpsc::SendError(Err(USimpleError::new(
+                    0,
+                    "Error already handled",
+                )))));
+            }
+        }
+    } else {
+        // This is the initial directory - try regular Stat::new first, then fallback to DirFd
+        match Stat::new(path, None, options) {
+            Ok(s) => s,
+            Err(_e) => {
+                // Try using our new DirFd method for the root directory
+                match DirFd::open(path) {
+                    Ok(dir_fd) => match Stat::new_from_dirfd(&dir_fd, path, options) {
+                        Ok(s) => s,
+                        Err(e) => {
+                            let error = e.map_err_context(
+                                || translate!("du-error-cannot-access", "path" => path.quote()),
+                            );
+                            if let Err(send_error) = print_tx.send(Err(error)) {
+                                return Err(Box::new(send_error));
+                            }
+                            return Err(Box::new(mpsc::SendError(Err(USimpleError::new(
+                                0,
+                                "Error already handled",
+                            )))));
+                        }
+                    },
+                    Err(e) => {
+                        let error = e.map_err_context(
+                            || translate!("du-error-cannot-access", "path" => path.quote()),
+                        );
+                        if let Err(send_error) = print_tx.send(Err(error)) {
+                            return Err(Box::new(send_error));
+                        }
+                        return Err(Box::new(mpsc::SendError(Err(USimpleError::new(
+                            0,
+                            "Error already handled",
+                        )))));
+                    }
```

Comment:
You use the same code in both error cases and so it might make sense to use a closure or function to remove the code duplication.